### PR TITLE
Added support for unwanted function; added more tests.

### PR DESCRIPTION
**Pull request by bwplotka**
Fixes: https://github.com/fatih/faillint/issues/7

## Changes

* No more globals :heart: 
* Separate tests for parsing (as it gets a bit more complex now) and tests.
* Add some test edge cases for imports as well (e.g named imports).
* Fixes some edge cases like whitespaces etc.
* Support for unwanted functions with `{Func1,Func2}` notion.

### CODE REVIEW

**Comment by bwplotka** [comment]
Damn, I was a minute too late

**Comment by fatih** [comment]
> Damn, I was a minute too late

Sorry about that but I've merged so you don't have to deal with the global stuff as @mweb already worked on that. Thanks again opening this PR, I'll check it soon :)

**Comment by bwplotka** [comment]
Should be fine now... Unless you will merge something straight to master again ;p Please use PullRequests - it's much more expected for the contributor to brace for conflicts ^^

**Comment by fatih** [comment]
> Should be fine now... Unless you will merge something straight to master again ;p Please use PullRequests - it's much more expected for the contributor to brace for conflicts ^^

It was a single line that fixed a string. It shouldn't cause any issues. Did it cause conflicts for you? I don't open PR's for small things like this in my personal projects as I think they don't require any review. And they are usually followup changes. Opening a PR for that string change is just to much burden for a indie developer who has several projects :) 

Otherwise I agree with you. I never push to Master, at my work or any other project that is big enough that warrants changes to be made via PR's. Thanks for bringing this to attention though.

**Comment by bwplotka** [comment]
Sure @fatih I am not going to force anything on your personal project, it's all up to you! ((: 

And yes somehow I got some minor conflict, but not a big deal! (: 

> Opening a PR for that string change is just to much burden for a indie developer who has several projects :) 

Maybe, if you are the alone author (but even then I think I disagree). As soon as you have more contributors than yourself I would always use PRs even on my personal projects. With PR:
* All commits on the master are buildable and fully tested by CI (!). Even by single line change, even a pro programmer can make typo etc, no one is perfect. And I guess you want your master to be stable.
* It helps to notify other contributors that some change is going, and give some time for a review... and maybe small coordination (: Since you start to have more contributors you might have some help in reviewing (I am happy to help, the codebase is small|). It's always good to have second pair of eyes.


Anyway, those are just suggestions, it's all up to you. Thanks for this project, and thanks for being helpful in accepting feedback and PRs :heart:  :+1:

Have a nice weekend! Hope this PR makes sense as it helps us a lot!

cc @pracucci for help in reviewing :hugs:

**Comment by fatih** [comment]
@bwplotka you make good points and I definitely don't undervalue them. Thanks for bringing them up. My intention wasn't to tell you to stop questionings things, I was just stating a fact how I operate nowadays. I had a big burnout (see: https://arslan.io/2018/10/09/taking-an-indefinite-sabbatical-from-my-projects/) and since then I try to keep things on my end simple to avoid that again.

I know pushing to master is a bad practice of course, and I don't do that at all in my professional work environment, but I've crossed the edge that that small details like this, until it really hurts people productivity, is something that can be neglected for my small projects. 

Once things get bigger and have more exposure, you're right. I know how things work at that scale (see: vim-go or structs package as an example).  Again it doesn't mean it's the correct way of doing things, I was just explaining how I operate. The points you bring up are all things people should follow :) 

> Have a nice weekend! Hope this PR makes sense as it helps us a lot!

It will definitely. Thanks again for working on this and also adapting it in other projects. I'm also planning to use it internally at GitHub for our services so it's great that it's open source and can be used widely in many places. I'll plan to review it soon.  Have a nice weekend as well.

File: faillint/faillint.go [hunk 1]
```diff
@@ -5,134 +5,176 @@ package faillint
...
 	ignoretests bool   // -ignore-tests flag
 }
 
-// Analyzer global instance of the linter (if possible use NewAnalyzer)
-var Analyzer = NewAnalyzer()
-
-// NewAnalyzer create a faillint analyzer
-func NewAnalyzer() *analysis.Analyzer {
+// New create a faillint analyzer.
+func New() *analysis.Analyzer {
```
Comment (fatih) [inline faillint/faillint.go]: I like NewAnalyzer better, would be happy if you can revert the naming.

File: faillint/faillint.go [hunk 2]
```diff
@@ -5,134 +5,176 @@ package faillint
...
+	"regexp"
 	"strconv"
 	"strings"
+	"unicode"
 
 	"golang.org/x/tools/go/analysis"
 )
 
+var (
+	pathRegexp = regexp.MustCompile(`(?P<import>[\w/.-]+[\w])(\.?{(?P<functions>[\w-,]+)}|)(=(?P<suggestion>[\w/.-]+[\w](\.?{[\w-,]+}|))|)`)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply faillint/faillint.go]: Good point, will do

File: faillint/faillint.go [hunk 3]
```diff
@@ -5,134 +5,176 @@ package faillint
...
+)
+
 type faillint struct {
 	paths       string // -paths flag
 	ignoretests bool   // -ignore-tests flag
 }
 
-// Analyzer global instance of the linter (if possible use NewAnalyzer)
-var Analyzer = NewAnalyzer()
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply faillint/faillint.go]: Totally see your point. I think there is no harm in getting this back as deprecated var.. (: and remove in v2

File: faillint/faillint.go (hunk 1 above)
Comment (bwplotka) [reply faillint/faillint.go]: Why? To me it's like including the type in variable name... (: `mInt` `someVarString` .. we used to do that 20 years ago when there were no modern IDEs.. now I think it's just a noise. You can check return type to learn about that it's analyser ((: Wdyt? What would be the argument to have Analyser in name?

File: faillint/faillint.go [hunk 4]
```diff
@@ -5,134 +5,176 @@ package faillint
...
 	a := &analysis.Analyzer{
 		Name:             "faillint",
-		Doc:              "report unwanted import path usages",
+		Doc:              "Report unwanted import path, and function usages",
 		Run:              f.run,
 		RunDespiteErrors: true,
 	}
 
-	a.Flags.StringVar(&f.paths, "paths", "", "import paths to fail")
+	a.Flags.StringVar(&f.paths, "paths", "", `import paths, functions or methods to fail.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply faillint/faillint.go]: Happy with that. Now is the question if declarations will be actually blocked as well .. will add to the tests 🤗

File: faillint/faillint.go [hunk 5]
```diff
@@ -5,134 +5,176 @@ package faillint
...
+E.g: foo,github.com/foo/bar,github.com/foo/bar/foo.{A}=github.com/foo/bar/bar.{C},github.com/foo/bar/foo.{D,C}`)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply faillint/faillint.go]: Good idea!

File: faillint/faillint.go [hunk 6]
```diff
@@ -5,134 +5,176 @@ package faillint
...
+	fn   []string
+	sugg string
+}
 
-		imp := imps[0]
-		suggest := ""
-		if len(imps) == 2 {
-			suggest = imps[1]
+func parsePaths(paths string) []path {
+	pathGroups := pathRegexp.FindAllStringSubmatch(trimAllWhitespaces(paths), -1)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply faillint/faillint.go]: It does not trim spaces in between, only the outrim spaces.

**Comment by bwplotka** [comment]
All addressed (:

File: README.md [hunk 7]
```diff
@@ -22,37 +23,42 @@ go get github.com/fatih/faillint
 `faillint` works on a file, directory or a Go package:
 
 ```sh
-$ faillint -paths "errors" foo.go # pass a file
-$ faillint -paths "errors" ./...  # recursively analyze all files
-$ faillint -paths "errors" github.com/fatih/gomodifytags # or pass a package
+$ faillint -paths "errors,fmt.{Errof}" foo.go # pass a file
+$ faillint -paths "errors,fmt.{Errof}" ./...  # recursively analyze all files
+$ faillint -paths "errors,fmt.{Errof}" github.com/fatih/gomodifytags # or pass a package
```
Comment (nnutter) [inline README.md]:
```suggestion
$ faillint -paths "errors,fmt.{Errorf}" foo.go # pass a file
$ faillint -paths "errors,fmt.{Errorf}" ./...  # recursively analyze all files
$ faillint -paths "errors,fmt.{Errorf}" github.com/fatih/gomodifytags # or pass a package
```

Comment (fatih) [reply README.md]: Good catch 👍

Comment (bwplotka) [reply README.md]: Ups, thanks!

File: faillint/faillint.go (hunk 1 above)
Comment (fatih) [reply faillint/faillint.go]: Just using `New()` doesn't explain immediately what it is returning. However `NewAnalyzer()` means exactly what it is. You're creating a new analzyer. You can also see some examples in the stdlib. For example from the `bufio` package:

NewReader()
NewBuffer()
NewWriter

Or from the `flag` package:

NewFlagSet()

There are many examples like this. You could argue that `errors` has a `New()` function, but that's a little bit different than these.

There it would be repetitive to say `errors.NewError()` and it's used in every single return expression as well, such as `return nil, errors.New()`. So keeping it short is important.

But for faillint, or any other package that provide a certain functionality, I would like to see the name of _what_ it is creating in the `New...` name.

**Comment by bwplotka** [comment]
Introduce this in Thanos already: https://github.com/thanos-io/thanos/pull/2228

**Comment by fatih** [comment]
That's great to see :) I'm ready to merge it but I still think the constructor should be reverted to `NewAnalysis()`. And then we can cut a new release (`v1.2.0`) so it's ready for consumption.

File: faillint/faillint.go (hunk 1 above)
Comment (bwplotka) [reply faillint/faillint.go]: Fair, but currently we have only one constructor, so if we would follow [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) rule, `New` is just enough. (: 

Let me know if that convinces you. If not, I will disagree but revert back and commit :smile:

Comment (fatih) [reply faillint/faillint.go]: You're hard to convince I have to admit it :) 

So yeah I still thing `NewAnalyzer()` is better than `New()` for this _package_. This is used only in one place, in our cmd/main.go file, but I want it to be descriptive. I don't think `New()` is descriptive enough. Thank you for understanding, I appreciate your co-operation .

Comment (bwplotka) [reply faillint/faillint.go]: Cool, Reverting :+1:

Comment (bwplotka) [reply faillint/faillint.go]: Done.

**Comment by bwplotka** [comment]
Thanks for a super quick review! :hugs:
